src/khepri_projection.erl

```diff
@@ -0,0 +1,282 @@
+%%
+%% Copyright © 2022 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+%% @doc Khepri projections
+%%
+%% Projections build a replicated ETS table using nodes from the store which
+%% match a {@link khepri_path:pattern()}. When a node matching a projection's
+%% pattern is changed in the store, the node is passed through the projection's
+%% {@link projection_fun()} to create record(s). These records are then stored
+%% in the projection's ETS table for all members in a Khepri cluster.
+%%
+%% Projections provide a way to query the store as fast as possible and are
+%% appropriate for lookups which require low latency and/or high throughput.
+%% Projection tables contain all records matching the pattern, though, so the
+%% memory footprint of a projection table grows with the number of nodes in
+%% the store matching the pattern.
+%%
+%% Projection ETS tables are owned by the Khepri cluster and are deleted when
+%% the cluster stops.
+%%
+%% Updates to projection tables are immediately consistent for the member of
+%% the cluster on which the change to the store is made and the leader member
+%% but are eventually consistent for all other followers.
+
+-module(khepri_projection).
+
+-include_lib("kernel/include/logger.hrl").
+
+-include("khepri_machine.hrl").
+-include("khepri_fun.hrl").
+
+-export([new/2, new/3, name/1]).
+
+ %% For internal use only
+-export([init/1, trigger/4]).
+
+-export_type([projection/0,
+              projection_fun/0,
+              options/0]).
+
+-record(?MODULE,
+        {%% The name of the projection. This atom should be unique to a node
+         %% as it will become the name of the ETS table.
+         name :: atom(),
+         %% A function compiled by `khepri_fun' which "projects" entries from
+         %% a Khepri store into records stored in a "projected" ETS table.
+         project_fun :: khepri_fun:standalone_fun(),
+         %% Options passed to `ets:new/2'
+         ets_options :: [atom() | tuple()]}).
+
+-opaque projection() :: #?MODULE{}.
+
+-type simple_projection_fun() ::
+fun((Path :: khepri_path:native_path(),
+     Payload :: khepri:data()) -> Record :: tuple()).
+%% A simple projection function.
+%%
+%% Simple projection functions only take the path and payload for a node in the
+%% store. The record produced by the function is used to create and delete
+%% objects in the ETS table.
+%%
+%% This function is compiled the same way as a transaction function: all
+%% side-effects are not allowed.
+
+-type extended_projection_fun() ::
+fun((Table :: ets:tid(),
+     Path :: khepri_path:native_path(),
+     OldPayload :: khepri:node_props(),
+     NewPayload :: khepri:node_props()) -> any()).
+%% An extended projection function.
+%%
+%% In some cases, a node in the store might correspond to many objects in a
+%% projection table. Extended projection functions are allowed to call ETS
+%% functions directly in order to build the projection table.
```

Comment:
We could specify that the return value of the extended projection function is ignored.